File: tests/test_eval_cache.py
from types import SimpleNamespace

import pandas as pd

from dssatcalibrator import orchestrator
from dssatcalibrator.spaces import ParameterSpace
from dssatcalibrator.spawn import SpawnResult


def _cfg(tmp_path):
    exe = tmp_path / "dscsm048"
    exe.write_text("fake exe", encoding="utf-8")
    return {
        "calibrator": {
            "name": "cache_test",
            "workdir": str(tmp_path / "work"),
            "dssat_dir": str(tmp_path),
            "dssat_exe": str(exe),
            "num_cores": 1,
            "cache_evaluations": True,
        },
        "source": {"hemp_dir": str(tmp_path)},
        "experiments": ["E1"],
        "parameters": {
            "genetic_cultivar": {
                "P1": {"active": True, "min": 0, "max": 10, "start": 5}
            }
        },
        "crops": [{"code": "HM"}],
        "engine": {"scalar_outputs": {"HWAM": "HWAM"}},
        "objective": {"weighting": "unified", "weights": {}, "error_model": {}},
    }


def _setup(cfg, tmp_path):
    space = ParameterSpace.from_config(cfg)
    crop = {"code": "HM", "filex_ext": "HMX", "genotype_stem": "HMGRO048"}
    obs = SimpleNamespace(table=pd.DataFrame())
    return (
        space,
        crop,
        tmp_path / "dscsm048",
        space.specs,
        tmp_path / "runs",
        obs,
        ["E1"],
        {"E1": [1]},
    )


def test_evaluate_thetas_persistent_cache_dedupes_and_reuses(tmp_path, monkeypatch):
    cfg = _cfg(tmp_path)
    setup = _setup(cfg, tmp_path)
    calls = {"jobs": 0}

    def fake_run_many(jobs, n_workers, on_done=None, warmup=0):
        calls["jobs"] += len(jobs)
        out = []
        for job in jobs:
            theta = dict(job["theta"])
            ev = pd.DataFrame({
                "treatment": [1],
                "variable": ["HWAM"],
                "sim": [100.0 + float(theta["P1"])],
                "meas": [100.0],
            })
            res = SpawnResult("success", tmp_path / "run", theta, evaluate=ev)
            out.append(res)
            if on_done:
                on_done(res)
        return out

    monkeypatch.setattr(orchestrator, "run_many", fake_run_many)

    first, setup = orchestrator.evaluate_thetas(
        cfg, [{"P1": 1.0}, {"P1": 1.0}, {"P1": 2.0}], setup=setup, n_workers=1
    )
    assert calls["jobs"] == 2
    assert [r.score for r in first] == [first[0].score, first[0].score, first[2].score]

    second, _ = orchestrator.evaluate_thetas(
        cfg, [{"P1": 1.0}, {"P1": 2.0}], setup=setup, n_workers=1
    )
    assert calls["jobs"] == 2
    assert [r.score for r in second] == [first[0].score, first[2].score]
    assert list((tmp_path / "work" / "cache_test" / "evaluation_cache").glob("*/*.json"))


def test_evaluation_cache_salt_invalidates_persisted_result(tmp_path, monkeypatch):
    cfg = _cfg(tmp_path)
    setup = _setup(cfg, tmp_path)
    calls = {"jobs": 0}

    def fake_run_many(jobs, n_workers, on_done=None, warmup=0):
        calls["jobs"] += len(jobs)
        return [
            SpawnResult(
                "success",
                tmp_path / "run",
                dict(job["theta"]),
                evaluate=pd.DataFrame({
                    "treatment": [1],
                    "variable": ["HWAM"],
                    "sim": [101.0],
                    "meas": [100.0],
                }),
            )
            for job in jobs
        ]

    monkeypatch.setattr(orchestrator, "run_many", fake_run_many)
    orchestrator.evaluate_thetas(cfg, [{"P1": 1.0}], setup=setup, n_workers=1)
    orchestrator.evaluate_thetas(cfg, [{"P1": 1.0}], setup=setup, n_workers=1)
    assert calls["jobs"] == 1

    cfg2 = _cfg(tmp_path)
    cfg2["calibrator"]["evaluation_cache_salt"] = "fresh"
    setup2 = _setup(cfg2, tmp_path)
    orchestrator.evaluate_thetas(cfg2, [{"P1": 1.0}], setup=setup2, n_workers=1)
    assert calls["jobs"] == 2

File: tests/testthat/test_eval_cache.R
test_that("persistent evaluation cache round-trips objective_result objects", {
  tmp <- tempfile("eval_cache_roundtrip_")
  dir.create(tmp, recursive = TRUE)
  exe <- file.path(tmp, "dscsm048")
  writeLines("fake exe", exe)
  cfg <- list(
    calibrator = list(name = "cache_test", workdir = file.path(tmp, "work"),
                      dssat_dir = tmp, dssat_exe = exe, cache_evaluations = TRUE),
    source = list(hemp_dir = tmp),
    experiments = list("E1"),
    parameters = list(genetic_cultivar = list(
      P1 = list(active = TRUE, min = 0, max = 10, start = 5)
    )),
    crops = list(list(code = "HM")),
    engine = list(scalar_outputs = list(HWAM = "HWAM")),
    objective = list(weighting = "unified", weights = list(), error_model = list())
  )
  space <- parameter_space_from_config(cfg)
  cache <- .evaluation_cache_from_setup(
    cfg,
    crop = list(code = "HM", filex_ext = "HMX", genotype_stem = "HMGRO048"),
    specs = space$specs,
    experiments = c("E1"),
    treatments = list(E1 = 1L),
    obs_table = data.frame(),
    exe = exe
  )
  key <- .eval_cache_key(cache, list(P1 = 1), c("E1"))
  result <- structure(list(
    score = 0.25,
    loglik = -0.125,
    residuals = data.frame(exp_id = "E1", treatment = 1L, user_var = "HWAM",
                           resid = 1, stringsAsFactors = FALSE),
    per_var = list(HWAM = list(n = 1L, RMSE = 1)),
    per_exp_var = data.frame(exp_id = "E1", user_var = "HWAM", n = 1L,
                             stringsAsFactors = FALSE)
  ), class = "objective_result")

  expect_true(.eval_cache_put(cache, key, result))
  loaded <- .eval_cache_get(cache, key)
  expect_s3_class(loaded, "objective_result")
  expect_equal(loaded$score, result$score)
  expect_equal(loaded$residuals$resid, 1)

  result2 <- result
  result2$score <- 0.125
  expect_true(.eval_cache_put(cache, key, result2))
  loaded2 <- .eval_cache_get(cache, key)
  expect_equal(loaded2$score, result2$score)
})

test_that("persistent evaluation cache dedupes and reuses objective results", {
  runner_env <- environment(evaluate_thetas)
  skip_if(bindingIsLocked("run_many", runner_env),
          "run_many binding is locked; exercised by source-based parity runner")

  tmp <- tempfile("eval_cache_")
  dir.create(tmp, recursive = TRUE)
  exe <- file.path(tmp, "dscsm048")
  writeLines("fake exe", exe)

  cfg <- list(
    calibrator = list(name = "cache_test", workdir = file.path(tmp, "work"),
                      dssat_dir = tmp, dssat_exe = exe, num_cores = 1L,
                      cache_evaluations = TRUE),
    source = list(hemp_dir = tmp),
    experiments = list("E1"),
    parameters = list(genetic_cultivar = list(
      P1 = list(active = TRUE, min = 0, max = 10, start = 5)
    )),
    crops = list(list(code = "HM")),
    engine = list(scalar_outputs = list(HWAM = "HWAM")),
    objective = list(weighting = "unified", weights = list(), error_model = list())
  )
  space <- parameter_space_from_config(cfg)
  setup <- list(
    space = space,
    crop = list(code = "HM", filex_ext = "HMX", genotype_stem = "HMGRO048"),
    exe = exe,
    specs = space$specs,
    run_root = file.path(tmp, "runs"),
    obs = observations(data.frame()),
    experiments = c("E1"),
    treatments = list(E1 = 1L)
  )

  old_run_many <- get("run_many", envir = runner_env)
  calls <- 0L
  fake_run_many <- function(jobs, n_workers, on_done = NULL, warmup = 0) {
    calls <<- calls + length(jobs)
    lapply(jobs, function(job) {
      theta <- job$theta
      spawn_result(
        "success",
        tmp,
        theta,
        evaluate = data.frame(
          treatment = 1L,
          variable = "HWAM",
          sim = 100 + as.numeric(theta$P1),
          meas = 100,
          stringsAsFactors = FALSE
        )
      )
    })
  }
  assign("run_many", fake_run_many, envir = runner_env)
  on.exit(assign("run_many", old_run_many, envir = runner_env), add = TRUE)

  first <- evaluate_thetas(cfg, list(list(P1 = 1), list(P1 = 1), list(P1 = 2)),
                           setup = setup, n_workers = 1L)$results
  expect_equal(calls, 2L)
  expect_equal(first[[1]]$score, first[[2]]$score)

  second <- evaluate_thetas(cfg, list(list(P1 = 1), list(P1 = 2)),
                            setup = setup, n_workers = 1L)$results
  expect_equal(calls, 2L)
  expect_equal(second[[1]]$score, first[[1]]$score)
  expect_equal(second[[2]]$score, first[[3]]$score)

  cache_files <- list.files(file.path(tmp, "work", "cache_test", "evaluation_cache"),
                            pattern = "\\.rds$", recursive = TRUE)
  expect_gt(length(cache_files), 0)
})
